File: 백준/Silver/2839. 설탕 배달/설탕 배달.py
N = int(input())
dp = [-1] * (5001)
dp[3] = 1
dp[4] = -1
dp[5] = 1
dp[6] = 2
dp[7] = -1

for i in range(8, 5000+1):
    prev_3 = dp[i-3]
    prev_5 = dp[i-5]
    
    if prev_3 != -1 and prev_5 != -1:
        dp[i] = min(prev_3 + 1, prev_5 + 1)
    elif prev_3 == -1 and prev_5 != -1:
        dp[i] = prev_5 + 1
    elif prev_3 != -1 and prev_5 == -1:
        dp[i] = prev_3 + 1
        
print(dp[N])

File: 백준/Silver/2839. 설탕 배달/설탕 배달.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStreamReader;

public class Main {
	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		
		boolean find = false;
		int result = 1666;
		
		for (int i = (N/5); i >= 0; i--) {
			int bongCnt = i;
			
			int fiveTotal = i * 5;
			int threeTotal = N - fiveTotal;
			
			if (threeTotal % 3 == 0) {
				bongCnt += (threeTotal / 3);
				find = true;
				result = Math.min(result, bongCnt);
			} else {
				continue;
			}
		}
		
		if (find) {
			System.out.println(result);
		} else {
			System.out.println(-1);
		}
	}
}
